### ISSUE TITLE
Round instead of floor percentage grades in PL interface

### ISSUE THREAD

**Issue by joelostblom** [PrairieLearn/PrairieLearn#10905]
Minor issue, but I had a couple of students confused by that their grades were floored down rather then rounded to the nearest integer. I explained that the exact percentage is kept in the exported grades, but it might be worthwhile also rounding these in the PL interface for clarity? (I personally round up on .5, but don't have a strong opinion on that for what is shown in the interface.)

![Image](https://github.com/user-attachments/assets/20dbc2b2-f1cd-4b84-8a77-25cc5b8b2758)

**Comment by mwest1066** [comment]
In general our policy is to floor rather than rounding so that students are not mislead into thinking that they have more points than they really do. For example, our thinking is that if a student has 99.99% then we don't want to display 100, because the student may be able to still work to actually get the additional points.

Displaying more decimal places is one way to work around this, as @ZacWarham is suggesting. Of course this makes the UI more cluttered, but that might be fine.

**Comment by ZacWarham** [comment]
> UI more cluttered

I did consider adding a `?` pop-up next to the scorebar that explains the rounding and/or shows the exact number but I thought this would be too cluttered on assessment pages where there is a scorebar for every question

> care about two decimal places

Personally it doesn't bother me, 1 is definitely cleaner than 2 - I opted for consistency with the points being 2dp but I'm happy to switch the PR to 1dp if its preferred

**Comment by mwest1066** [comment]
(unifying discussion here from this issue and #10909)

From https://github.com/PrairieLearn/PrairieLearn/issues/10905#issuecomment-2466308385:

> I did consider adding a `?` pop-up next to the scorebar that explains the rounding and/or shows the exact number but I thought this would be too cluttered on assessment pages where there is a scorebar for every question

I like the idea of a `?` popover! See below for a proposal using this idea.

> Personally it doesn't bother me, 1 is definitely cleaner than 2 - I opted for consistency with the points being 2dp but I'm happy to switch the PR to 1dp if its preferred

This is certainly an initial reasonable assumption. However, I'm also not convinced that 2dp is necessary for points, although at least for points there is the justification that points are often smaller numbers than percentages (e.g., the max points for a question might be 5 or 10, so if we think that one part in a thousand is the correct precision then 2dp would be correct).

From https://github.com/PrairieLearn/PrairieLearn/pull/10909#discussion_r1835492737:

> I wouldn't use toFixed because it forces zeros to the right (54.00). You can use formatPoints instead.

I agree that we want to trim trailing zeros when displaying points, but when we are displaying scores I think we might actually want to pick a precision and always use it. Otherwise "54" could mean any of "53.6 rounded up", "54.9 floored", "54.00 accurate to 2dp", or some other combination of rounding and precision.

**Comment by mwest1066** [comment]
## Proposal: popovers for score details

Overall, I agree with the original point of this issue that we should give students access to more detailed information about their scores. However, I think that showing additional precision on all scorebars will clutter the UI and is generally not necessary. Building on @ZacWarham's idea in https://github.com/PrairieLearn/PrairieLearn/issues/10905#issuecomment-2466308385, I propose:
* On pages where we are showing a single scorebar (e.g., assessment instance pages), we should show the scorebar exactly as we do now (floored to 0dp) but add a `?` popover that shows the exact score and explains the fact that this is floored. I'd suggest adding an optional `options` parameter to `Scorebar()` so that we can write `Scorebar(score_perc, {detailsPopover: "instructor"})` or something like that. The `detailsPopover` value could be:
    * **`undefined`:** no popover.
    * **`"instructor"`:** shows the exact value to 4dp with trailing zeros using `toFixed()`, explains the rounding (say "rounded down" rather than "floored"), says that students will see "54" listed on overview pages and they can also click `?` on individual assessment pages to see their exact scores, and explains the justification (students will never be shown more points then they actually have).
    * **`"student"`:** shows the exact value to 4dp with trailing zeros, explains the rounding, explains the justification that you will never be shown more points than you actually have.
* When we show many small scorebars together (e.g., on overview pages like Assessments) then we should continue to show the scorebars exactly as they are now (floored to 0dp, no popover). For example:
    <img width="108" alt="image" src="https://github.com/user-attachments/assets/ac3acdc1-1f6e-418d-9bb6-1b610ce32dca">
* In columns like the one shown above, we should add a `?` popover to the `<th>` table heading cell with a popover that explains the rounding, explains the justification, and says where to find more detail (click through into individual assessment or assessment instance pages. I suggest adding a `scorebarHeaderPopover(type)` function that shows the `?` button, where `type` is either `"instructor"` or `"student"`, similar to the above.
* On the `instructorAssessmentStatistics` page we could add popovers to show exact values for the summary statistics (e.g., [here](https://github.com/PrairieLearn/PrairieLearn/blob/e0f10ee96084f68747142c9c104645e2508b72d2/apps/prairielearn/src/pages/instructorAssessmentStatistics/instructorAssessmentStatistics.html.ts#L112)) although these are already available in the CSV download so this is less critical.

## Questions

1. Rather than showing all scorebars as floored to 0dp, should we show them all as floored to 1dp? I think that more precision than this is not useful, but I can see that sometimes 89.8 is importantly different to 89. I personally think that the extra precision of 1dp is not worth the UI clutter, but I'd like to hear other opinions here.
2. Should we sometimes round to nearest integer rather than flooring? For student scores I think that flooring is the way to go, because who knows what policy the instructor is going to impose outside of PL and I don't want to show a student that they got 90, which could be the cutoff for an A grade, when they actually got 89.8 and thus only got a B. But for instructor statistics like the average assessment score it is actually probably better to round rather than floor. I tend to think we should do this. If we do it, then I think we should add more mandatory options to `Scorebar()` where we have to indicate whether this is an individual score (which will be floored) or an average score (which will be rounded).

**Comment by jonatanschroeder** [comment]
Just adding another reason to floor instead of round: at the end of term students will often calculate their final course grade based on the scores they see. Some students contact me asking me to, for example, round their grades up for them to get to the next letter grade, but when I check my calculations (which use the full gradebook with non-rounded assessment grades) I see is already at the next letter grade, mostly due to the accumulated effect of adding up the fractions of grades in multiple assessments (a simple example: three assessments at 89.6, 89.6 and 90.8 would give an average of 90, but would appear to be just below it by rounding down each assessment). I prefer this than the opposite, i.e., students assuming they are at the next letter grade because of accumulated rounding ups and then get confused when their grade is lower (e.g., three assessments at 89.5).

**Comment by nwalters512** [comment]
Because I think it can be helpful to see if others have already solved this problem: what do Canvas/Moodle/Gradescope/etc. do for fractional points/percentages?

**Comment by mwest1066** [comment]
> Because I think it can be helpful to see if others have already solved this problem: what do Canvas/Moodle/Gradescope/etc. do for fractional points/percentages?

Gradescope: I'm not sure exactly what it's doing, it seems to show points to at least 1dp with trailing zeros, but some points are shown to 2dp. It might be using fixed point arithmetic for this and always showing full precision? Summary point statistics are shown to 2dp. It doesn't seem to show percentages at all.

Canvas: the gradebook shows 2dp without trailing zeros, and it appears to be rounding (not flooring), although it's not quite clear to me whether the rounding is happening in Canvas or Gradescope. Percentage statistics are shown to 2dp.

**Comment by joelostblom** [comment]
Thanks for sharing these great suggestions!

> If we were to display more decimal places (floored), would one be enough? Does anyone care about two decimal places?

I vote that 1dp is enough.

> Rather than showing all scorebars as floored to 0dp, should we show them all as floored to 1dp?

I am slightly in favor of this, but it would depend on how cluttered it looks. As long as students can see details on their score easily, I don't think it matters that much if the displayed number is rounded to an integer or 1dp.

**Comment by nwalters512** [comment]
Some thoughts in no particular order:

- I agree that we don't want to make this configurable (at least not yet).
- I agree that one decimal place would probably be sufficient.
- I really would rather not throw even more popovers at our various problems (if we can avoid it).
- Rather than popovers, could we use a tooltip and show the full-precision value there? This would avoid additional clutter, though it'd be harder to discover.
- Rather than any of this, could we just give students access to a raw, full-precision CSV export from the gradebook page? This is also not particularly discoverable, but gives instructors a straightforward way to answer things like "my exam says 89.5% but I need an 89.57% for an A???" and also allows power users to easily get their grades into Excel/Pandas/whatever.

**Comment by jonatanschroeder** [comment]
> * Rather than popovers, could we use a tooltip and show the full-precision value there? This would avoid additional clutter, though it'd be harder to discover.

I like the tooltip idea, with the usual accessibility caveats.

> * Rather than any of this, could we just give students access to a raw, full-precision CSV export from the gradebook page? This is also not particularly discoverable, but gives instructors a straightforward way to answer things like "my exam says 89.5% but I need an 89.57% for an A???" and also allows power users to easily get their grades into Excel/Pandas/whatever.

That's not a bad idea either.

**Comment by nwalters512** [comment]
> with the usual accessibility caveats

Yeah, this is tricky. Making this accessible would require making all the scorebars tab stops, which IMO could hurt usability on pages where we show a lot of them. Bootstrap also doesn't let you select information from tooltips, so someone looking to export a full-precision number would have to retype if by hand.

**Comment by nwalters512** [comment]
In the interest of being able to do something about #10909 (and ultimately close this issue), I'd like to propose that we do the following:

- Add full-precision grade exports to the student gradebook in CSV form.
- *Maybe* rework #10909 to round to one decimal place in all scorebars. If students have access to full-precision data somewhere, I'm not totally convinced it's worth cluttering every single scorebar on the site with an extra decimal; I'd be happy to just close that PR once we have student-facing CSV exports.

**Comment by nwalters512** [comment]
Let's start with CSV exports for students; we can update the score bars later if needed.
